### PR TITLE
Basic Media Controller support (+ media pipeline configuration)

### PR DESCRIPTION
**Pull request by otyshchenko1**
I would be ok to squash into single patch if needed.

### CODE REVIEW

**Comment by otyshchenko1** [comment]
Corresponding PR https://github.com/xen-troops/meta-xt-prod-devel/pull/253

**Comment by otyshchenko1** [comment]
Added https://github.com/xen-troops/camera_be/pull/9/commits/9e50b1517cae0ddb365522ee3c25bdf99ce64c93

**Comment by otyshchenko1** [comment]
Tested HDMI_IN & CVBS with DomU

**Comment by iusyk** [comment]
1. mMediaController - what is the reason for the shared_ptr ? (unique_ptr fits better)
2. CameraHandler::CameraHandler has the following 
    mMediaController.reset(new MediaController(mediaId, config));
    ctor is for construction, it is clear that mMediaController is not initialized, why to we need the call of ;reset' ?

3. Config(const std::string& fileName); , 
  - explicit is required
   - better to have the following declaration Config(std::string fileName), in this case, compiler may support the move semantic

4. is it supposed to use 
   - copy ctor
  - move ctor
  - operator=
    ?
   if not -add the corresponded declaration

5. MediaController::MediaController(const std::string devName, ConfigPtr config):
    -remove const - not required but does not allow using of move  
    - config is copied , but there are not a corresponded ctor, does it mean that default one has to be used ? potentially - this is a problem

6. looks like classes are not inheritable , please use 'final' specifier

7. Config::PipelineConfig config = mConfig->getPipelineConfig();
    - may be it is reasonable to return const Config::PipelineConfig& from getPipelineConfig to avoid the creation of object copies.
  final code may look like:
  const auto& config= mConfig->;getPipelineConfig();

8. what is the sense of ret assignment  looking at fact that this is not initialized variable?
    if (!m_mediaDevice) {
        ret = -ENODEV;
        LOG(mLog, ERROR) << "Failed to open device " << mDevPath;
        goto out;
    }
   please, initialize the variable with invalid value
   int ret= -ENODEV;
9. label 'out'  - as for me, it is better forget about the goto 
as option we can use the following (of cause, it is possible to improve it using the place holder for the call and variadic templates)

auto throwIfBad= [&mLog, &mDevPath, &ret](bool res, std::string&& sMsg, std::string dMsg) {
        if(!res){
            LOG(mLog, ERROR) << sMsg << dMsg;
            throw Exception("Failed to initialize media device " + mDevPath, -ret);
        }
    };

samples of the usage
----------------------------------------
instead of 
if (!m_mediaDevice) {
        ret = -ENODEV;
        LOG(mLog, ERROR) << "Failed to open device " << mDevPath;
        goto out;
    }

we can do 

m_mediaDevice = media_device_new(mDevPath.c_str());
throwIfBad(m_mediaDevice != nullptr, "Failed to open device ",  mDevPath);

OR
---------------------------------------------

instead of 

ret = media_device_enumerate(m_mediaDevice);
    if (ret) {
        LOG(mLog, ERROR) << "Failed to enumerate device " << mDevPath;
        goto out;
    }

we can use
ret = media_device_enumerate(m_mediaDevice);
throwIfBad(0 == ret, "Failed to enumerate device ", mDevPath );

**Comment by arminn** [comment]
1)unique_ptr and make_unique are good choices
3) "better to have the following declaration Config(std::string fileName)," hmmm, but it looks like fileName won't be saved/moved to any object, maybe const& is better?

File: src/MediaController.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+}
+
+void MediaController::showInfo()
+{
+    const struct media_device_info *info = media_get_info(m_mediaDevice);
+    char version[16];
+
+    LOG(mLog, DEBUG) << "Media device information";
+
+    sprintf(version, "%u.%u.%u",
```
Comment (arminn) [inline src/MediaController.cpp]: snprintf ?

Comment (otyshchenko1) [reply src/MediaController.cpp]: good point

**Comment by otyshchenko1** [comment]
@iusyk @arminn thank you for the review. Please see 421c5a2eb8f8b4f1b992c56048bcc4ac13ca4036 where almost all comments were addressed (I hope). If you ok with proposed changes I will squash them with main patch.

**Comment by otyshchenko1** [comment]
Added https://github.com/xen-troops/camera_be/pull/9/commits/015238609934657a36108625e4b383311f3ddfa2

**Comment by otyshchenko1** [comment]
Update by squashing temp patch (where comments are addressed) with main patches.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -47,5 +50,7 @@ set(XENBE_LIB xenbe)
 target_link_libraries(${PROJECT_NAME}
 	${XENBE_LIB}
 	${V4L2_LIBRARY}
+	${MEDIACTL_LIBRARIES}
+	v4l2subdev
```
Comment (andr2000) [inline src/CMakeLists.txt:57]: Why this doesn't go into MEDIACTL_LIBRARIES? My understanding is that we didn't require this before and this is MEDIACTL_LIBRARIES which requires it

File: src/CameraHandler.cpp [hunk 3]
```diff
@@ -51,7 +54,50 @@ void CameraHandler::init(std::string uniqueId)
...
+    std::string videoId, mediaId;
+    parseUniqueId(uniqueId, videoId, mediaId);
+
+    if (!mediaId.empty()) {
+        LOG(mLog, DEBUG) << "media-id is not empty, media pipeline needs to be configured";
+
+        mMediaController = MediaControllerPtr(new MediaController(mediaId));
+    }
+
+    if (!videoId.empty())
```
Comment (andr2000) [inline src/CameraHandler.cpp:76]: Can it be that both videoID and mediaID configured by mistake?

File: src/CMakeLists.txt (hunk 2 above)
Comment (otyshchenko1) [reply src/CMakeLists.txt:57]: because this is not a part of MEDIACTL_LIBRARIES. Although provided by media-ctl, this is a separate library

Comment (andr2000) [reply src/CMakeLists.txt:57]: ok

**Comment by andr2000** [comment]
`Reviewed-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>`

File: src/CameraHandler.cpp (hunk 3 above)
Comment (otyshchenko1) [reply src/CameraHandler.cpp:76]: Theoretically yes, as everything in domain config, whether it is a display connector or block device number.  I have documented three possible options at the moment https://github.com/xen-troops/meta-xt-prod-devel/pull/253/commits/6081e0bd2e43b4170f5f0aa1afe8c4b6e49671d1#diff-2577518880b5e4b8f8d6b877d3042b5a for user to not dig into details too much what the proper configuration is, but choose needed camera type.

**Comment by otyshchenko1** [comment]
update by adding R-b

**Pull request merged**